appserver/web/war-util/src/test/java/org/glassfish/web/loader/WebappClassLoaderTest.java

```diff
@@ -0,0 +1,183 @@
+ * and therefore, elected the GPL Version 2 license, then the option applies
+ * only if the new code is made subject to such option by the copyright
+ * holder.
+ */
+package org.glassfish.web.loader;
+
+import static org.junit.Assert.assertTrue;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+
+import org.apache.naming.resources.FileDirContext;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class WebappClassLoaderTest {
+
+    private static final int EXECUTION_COUNT = 100;
+
+    private CountDownLatch latch;
+
+    private ExecutorService executor;
+    private File junitJarFile;
+
+    @Before
+    public void setup() throws URISyntaxException {
+        // Run 3 methods at the same time, and make the pool large enough to increase
+        // the chance of a race condition
+        executor = Executors.newFixedThreadPool(60);
+
+        // Require a minimum number of executions before completing
+        latch = new CountDownLatch(EXECUTION_COUNT);
+
+        // Fetch any JAR to use for classloading
+        junitJarFile = new File(Test.class.getProtectionDomain().getCodeSource().getLocation().toURI());
+    }
+
+    @After
+    public void shutdown() throws InterruptedException {
+        if (executor != null) {
+            executor.shutdownNow();
+            assertTrue("Executor could not shutdown. This could mean there is a deadlock.",
+                    executor.awaitTermination(10, TimeUnit.SECONDS));
+        }
+    }
+
+    @Test
+    public void check_findResourceInternalFromJars_thread_safety() throws Exception {
+        final WebappClassLoader webappClassLoader = new WebappClassLoader(getClass().getClassLoader(), null);
+        webappClassLoader.start();
+        webappClassLoader.setResources(new FileDirContext());
+
+        CompletableFuture<Void> result = new CompletableFuture<>();
+
+        // Create the tasks to run
+        Runnable lookupTask = waitAndDo(result, () -> lookup(webappClassLoader));
+        Runnable addTask = waitAndDo(result, () -> add(webappClassLoader));
+        Runnable closeTask = waitAndDo(result, () -> webappClassLoader.closeJARs(true));
+
+        try {
+            // Run the methods at the same time
+            for (int i = 0; i < EXECUTION_COUNT; i++) {
+                executor.execute(addTask);
+                executor.execute(lookupTask);
+                executor.execute(closeTask);
+            }
+
+            // Wait for tasks to execute
+            assertTrue("The tasks didn't finish in the allowed time.",
```

Comment:
It waits only for 100 tasks of the 300 submitted. Since 'add' and 'closeJar' are small/fast tasks, this will never be an issue (except in case of deadlock)

Reply:
Yep, this is there exclusively for deadlocks